docs/src/man/joins.md

```diff
@@ -241,12 +241,12 @@ julia> b = DataFrame(ID = [20, 60], Job = ["Lawyer", "Doctor"])
 
 julia> outerjoin(a, b, on=:ID, validate=(true, true), indicator=:source)
 3×4 DataFrame
-│ Row │ ID    │ Name    │ Job     │ source       │
-│     │ Int64 │ String⍰ │ String⍰ │ Categorical… │
-├─────┼───────┼─────────┼─────────┼──────────────┤
-│ 1   │ 20    │ John    │ Lawyer  │ both         │
-│ 2   │ 40    │ Jane    │ missing │ left_only    │
-│ 3   │ 60    │ missing │ Doctor  │ right_only   │
+│ Row │ ID    │ Name    │ Job     │ source     │
+│     │ Int64 │ String? │ String? │ Cat…       │
```

Comment:
It would be nice to use the available space here, as for other types, and print `Categoric…`.

Reply:
I can fix it but let us decide on the rules of abbreviation of `Categorical`, because this was the problem at the time when this was discussed originally.

Do we print `Cat`/`Categoric`/`Categorical` (or some other list) or allow to shorten `Categorical` at any letter?

Reply:
Sorry, my suggestion was clear. I meant that we should print `Cat…` when the column isn't wider than that, but use all the available space when there is (just like for other types).

Reply:
> but use all the available space when there is (just like for other types).

The issue is that we do not do this for other types now (that is why I asked). We either show the whole type or trim what is printed to up to 8 characters.

I have pushed a change that follows what you wanted if I understand your intention correctly (please comment if this is correct and then I will update the documentation).

Reply:
Woops, I realize I didn't even know the current rules, yet I was the one who proposed that one at https://github.com/JuliaData/DataFrames.jl/pull/1499#issuecomment-417861696! It's interesting to note that there are very few types whose names are above 8 characters, so in practice only `CategoricalValue` was in this case and we special-cased it.

Maybe showing `Cat` unless there's enough space to print `CategoricalValue` in full is OK then. Otherwise, I feel like we should find a more general rule that would apply to other types. Do we have other examples of types with long names? :-/

Reply:
Probably there are other types (e.g. if we store arrays of arrays/tuples/named tuples):
```
julia> DataFrame(a=(1,2),b=[[1]],c=(a=1,b=2))
1×3 DataFrame
│ Row │ a      │ b      │ c              │
│     │ Tuple… │ Array… │ NamedTup…      │
├─────┼────────┼────────┼────────────────┤
│ 1   │ (1, 2) │ [1]    │ (a = 1, b = 2) │
```
but 8 was chosen so that "typical types", even these above, were printed relatively OK.

So in practice I think only `CategoricalValue` is problematic. Given what you commented I think my current proposal to:
* print full type if it is possible (very rare)
* print "Categorical..." for wider columns
* print "Cat..." for columns not wide enough for "Categorical..."

should be OK. If you approve I will rewrite the docstrings to follow the current implementation (which does what I have described here).

Reply:
Fine with me. Though maybe a smarter algorithm would be to break before the last character that is not a lowercase letter and for which the content fits. That would give the behavior you describe for `CategoricalValue`, but also improve the printing of `NamedTuple` above so that you get `NamedTuple` if there's enough space.

Reply:
I would leave as is for now - as I prefer to close this PR and move forward with other that are waiting for it to be merged (what you describe is doable and better, but would require careful testing, the reason is that there are surprisingly many "corner cases" in names of types in Julia: `Union`s, parameters, numbers, curly braces etc. - I remember struggling with them last time we changed this). We can of course improve it in the future.

So I now will update the docstrings.

Reply:
Sure.

Reply:
Well - as usual :smile: - you have prompted me to fix things in the end. I have updated the `compacttype` implementation to do what I think is the right thing. That is:
* always try to print at least 8 text width **wide** type name (and correctly handle wide characters which we did not do earlier), also correctly always print 8 text width wide compacttype if it has to be truncated to 8 characters (this was also incorrect before in some cases)
* Handle `Any`, `Missing`, `Union` and `Categorical...` as special cases
* if type has to be compacted (i.e. it does not fit the `maxwidth`) then only try printing its name (so then we always drop the parameter part for `UnionAll` and concrete parametric types)

I did not do "truncate at first capital letter" as I came to the conclusion that it is better just to try printing as much as possible within `maxwidth` given the other rules above (with the exception of categorical which is abbreviated to `Cat` or `Categorical` as described in the rules above).